File: cpp/15_反转链表.cpp
#include <iostream>
#include <vector>

typedef struct  ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
} ListNode;

class Solution {
public:
    // 使用向量来初始化空链表
    void init_use_vec(ListNode* head, std::vector<int>& vec) {
        if (head == nullptr) {
            return;
        }

        ListNode* tmp = head;
        for (auto i : vec) {
            tmp->next = new ListNode(i);
            tmp = tmp->next;
        }
    }

    // 打印链表
    void print_ListNode(ListNode* head) {
        if (head == nullptr) {
            return;
        }

        ListNode* tmp = head;
        while (tmp) {
            std::cout << tmp->val << " ";
            tmp = tmp->next;
        }
        std::cout << std::endl;
    }

    // 反转链表
    ListNode* ReverseList(ListNode* head) {
        if (head == nullptr) {
            return nullptr;
        }

        ListNode* p = head;
        ListNode* q = head->next;
        head->next = nullptr;

        while (q) {
            ListNode* tmp = q->next;
            q->next = p;
            p = q;
            q = tmp;
        }

        return p;
    }
};

int main(int argc, char* argv[])
{
    Solution s;
    ListNode* head = new ListNode(0);
    std::vector<int> vec = {2, 3, 9, 5, 1, 7};

    s.init_use_vec(head, vec);
    s.print_ListNode(head);

    head = s.ReverseList(head);
    s.print_ListNode(head);

    return 0;
}

File: python/15_反转链表.py
#!bin/bash python3
#-*- coding:utf-8 -*-

import time
import ListNode

"""
class ListNode():
    def __init__(self, x=0):
        self.val = x
        self.next = None
"""

class Solution():
    def ReverseList1(self, head):
        """
        第一次的方式，迭代法
        """
        if head == None:
            return None

        p = head
        q = head.next
        head.next = None

        while q:
            tmp = q.next
            q.next = p
            p = q
            q = tmp
        
        return p
    
    def ReverseList2(self, head):
        """
        迭代法优化版
        """
        if head == None or head.next == None:
            return head
        
        p = head.next
        head.next = None

        while p != None:
            q = p.next
            p.next = head
            head = p
            p = q
        
        return head
    
    def ReverseList3(self, head):
        """
        递归法
        """
        if head == None or head.next == None:
            return head
        
        curr = self.ReverseList3(head.next)
        head.next.next = head
        head.next = None

        return curr

    def test(self, head, ifPrint=True):
        """
        测试函数
        """
        func_vec = [self.ReverseList1,
                    self.ReverseList2,
                    self.ReverseList3]
        print("=====")
        for func in func_vec:
            start = time.time()
            result = func(head)
            end = time.time()

            print("time(us): {:>5.2f}, result: ".format((end - start)*10**6), end="")
            if ifPrint:
                result.listPrint(False, " ")
            else:
                print("too many numbers to display")
            
            head = func(result) # 注意恢复
            

def main():
    vec = [2, 3, 9, 5, 1, 7]
    head = ListNode.ListNode(0)
    head.creatFromList(vec)

    vec2 = [0] * 500
    head2= ListNode.ListNode(0)
    head2.creatFromList(vec2)

    s = Solution()
    s.test(head.next, True)
    s.test(head2.next, False)


if __name__ == "__main__":
    main()